src/main/java/client/GameClient.java

```diff
@@ -0,0 +1,1307 @@
+      log("Initial connection attempt failed. Reason: " + reason + ".");
+      printToConsole("\nFailed to connect to server: " + reason);
+    }
+    // If oldState was already RECONNECTING or DISCONNECTED, further user messages might be
+    // redundant here.
+
+    this.currentSessionId = null;
+    this.availableCasesCache = null;
+    this.publicGamesCache = null;
+    this.hostPlayerIdInSession = null; // Clear session-specific data
+
+    if (channel != null && channel.isOpen()) {
+      try {
+        channel.close();
+      } catch (IOException e) {
+        logError("IOException closing channel on disconnect", null);
+      }
+    }
+    channel = null;
+
+    // Stop the listener thread if it's running
+    if (networkListenerThread != null && networkListenerThread.isAlive()) {
+      networkListenerThread.interrupt(); // Signal it to stop
+      try {
+        networkListenerThread.join(100); // Brief wait for it to die
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+    networkListenerThread = null;
+
+    // The run() loop will now handle the logic for RECONNECTING or displaying DISCONNECTED prompt.
+    // No explicit "Attempting to reconnect..." print here.
+  }
+
+  public void stopClient() {
+    log("Client stop requested.");
+    running.set(false);
+    if (networkListenerThread != null) networkListenerThread.interrupt();
+    if (channel != null && channel.isOpen()) {
+      try {
+        channel.close();
+      } catch (IOException e) {
+        logError("Exception closing channel on stop", null);
+      }
+    }
+    connected.set(false);
+    currentState.set(ClientState.EXITING);
+    printToConsole("Exiting client...");
+  }
+
+  private void shutdownClientResources() {
+    if (networkListenerThread != null) {
+      try {
+        networkListenerThread.join(500);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+    log("Client resources shut down.");
+  }
+
+  private void log(String message) {
+    System.out.println("[" + LocalTime.now().format(TIME_FORMATTER) + " CLIENT] " + message);
+  }
+
+  private void logError(String message, Throwable t) {
+    System.err.println("[" + LocalTime.now().format(TIME_FORMATTER) + " CLIENT ERROR] " + message);
+    if (t != null && !(t instanceof ConnectException)) {
+      t.printStackTrace(System.err);
+    }
+  }
+
+  private void printToConsole(String message) {
+    consoleLock.lock();
+    try {
+      System.out.println(message);
+    } finally {
+      consoleLock.unlock();
+    }
+  }
+
+  private void handleNpcMoved(NpcMovedDTO nmd) {
+    // You can make this message more subtle if desired, e.g., not starting with [GAME INFO]
+    // if it's considered part of normal world updates.
+    printToConsole("[GAME INFO] " + nmd.toString()); // Uses the DTO's helpful toString() method
+  }
+}
```

Comment:
you made a state machine, but really hard to read when it's all in a single file

try to make this more solid